File: server/internal/usecase/wish.go
package usecase

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/kotarowakuwaku/asset-wish/server/internal/domain"
)

type WishUsecase struct {
	tx           TxManager
	wishes       WishRepository
	accounts     AccountRepository
	transactions TransactionRepository
	now          Clock
	newID        IDGenerator
}

func NewWishUsecase(
	tx TxManager,
	wishes WishRepository,
	accounts AccountRepository,
	transactions TransactionRepository,
	now Clock,
	newID IDGenerator,
) *WishUsecase {
	return &WishUsecase{
		tx:           tx,
		wishes:       wishes,
		accounts:     accounts,
		transactions: transactions,
		now:          now,
		newID:        newID,
	}
}

func (u *WishUsecase) List(ctx context.Context, status *domain.WishStatus) ([]domain.Wish, error) {
	return u.wishes.List(ctx, status)
}

func (u *WishUsecase) Create(
	ctx context.Context,
	title string,
	amount domain.Money,
	category domain.WishCategory,
	priority int,
	deadline *time.Time,
) (domain.Wish, error) {
	w, err := domain.NewWish(u.newID(), title, amount, category, priority, deadline)
	if err != nil {
		return domain.Wish{}, err
	}
	if err := u.wishes.Create(ctx, w); err != nil {
		return domain.Wish{}, err
	}
	return w, nil
}

// UpdateWishInput は PATCH の部分更新を表す。nil の項目は変更しない。
//
// Status が無いのは、状態遷移を Commit / Pay / Drop に限るため（不変条件6）。
// ClearDeadline があるのは、Deadline が nil のときに「変更しない」と
// 「期限を外す」を区別できないため。
type UpdateWishInput struct {
	Title         *string
	Amount        *domain.Money
	Category      *domain.WishCategory
	Priority      *int
	Deadline      *time.Time
	ClearDeadline bool
}

func (u *WishUsecase) UpdateContent(ctx context.Context, id uuid.UUID, in UpdateWishInput) (domain.Wish, error) {
	w, err := u.wishes.Get(ctx, id)
	if err != nil {
		return domain.Wish{}, err
	}

	title := w.Title
	if in.Title != nil {
		title = *in.Title
	}
	amount := w.Amount
	if in.Amount != nil {
		amount = *in.Amount
	}
	category := w.Category
	if in.Category != nil {
		category = *in.Category
	}
	priority := w.Priority
	if in.Priority != nil {
		priority = *in.Priority
	}
	deadline := w.Deadline
	switch {
	case in.ClearDeadline:
		deadline = nil
	case in.Deadline != nil:
		deadline = in.Deadline
	}

	// 検証を通すためにコンストラクタを経由する。ここを素通りさせると、
	// 空文字のタイトルや 0 円のウィッシュが更新経由で入り込む。
	updated, err := domain.NewWish(w.ID, title, amount, category, priority, deadline)
	if err != nil {
		return domain.Wish{}, err
	}
	// NewWish は検討中で作るため、現在の状態を戻す。
	// 状態を動かすのは Commit / Pay / Drop だけ。
	updated.Status = w.Status

	if err := u.wishes.UpdateContent(ctx, updated); err != nil {
		return domain.Wish{}, err
	}
	return updated, nil
}

// Commit は 検討中 → 確定 に遷移させる。
// 確定した時点で実質資産から控除される（不変条件3）。
func (u *WishUsecase) Commit(ctx context.Context, id uuid.UUID) (domain.Wish, error) {
	return u.transit(ctx, id, (*domain.Wish).Commit)
}

// Drop は 見送り に遷移させる。
func (u *WishUsecase) Drop(ctx context.Context, id uuid.UUID) (domain.Wish, error) {
	return u.transit(ctx, id, (*domain.Wish).Drop)
}

// transit は取得・遷移・保存をまとめる。
//
// 遷移して良いかの判定は domain.Wish のメソッドが持つ。usecase は
// 「どの遷移を起こしたいか」だけを知る（不変条件6）。
func (u *WishUsecase) transit(
	ctx context.Context,
	id uuid.UUID,
	transition func(*domain.Wish) error,
) (domain.Wish, error) {
	var result domain.Wish

	err := u.tx.RunInTx(ctx, func(ctx context.Context) error {
		w, err := u.wishes.Get(ctx, id)
		if err != nil {
			return err
		}
		if err := transition(&w); err != nil {
			return err
		}
		if err := u.wishes.UpdateStatus(ctx, w); err != nil {
			return err
		}
		result = w
		return nil
	})
	if err != nil {
		return domain.Wish{}, err
	}
	return result, nil
}

// Pay は 確定 → 完了 に遷移させ、口座残高を減らして履歴を残す
// （detailed-design 3.2.3）。
//
// 支払い後、そのウィッシュは確定支出から外れ、同額だけ残高が減る。
// **実質資産は支払いの前後で変わらない。** これが正しい挙動になる。
func (u *WishUsecase) Pay(
	ctx context.Context,
	id uuid.UUID,
	accountID uuid.UUID,
	occurredOn time.Time,
) (domain.Wish, error) {
	var paid domain.Wish

	err := u.tx.RunInTx(ctx, func(ctx context.Context) error {
		w, err := u.wishes.Get(ctx, id)
		if err != nil {
			return err
		}
		if err := w.Pay(); err != nil {
			return err
		}
		if err := u.wishes.UpdateStatus(ctx, w); err != nil {
			return err
		}

		account, err := u.accounts.Get(ctx, accountID)
		if err != nil {
			return err
		}
		account.ApplyDelta(-w.Amount, u.now())
		if err := u.accounts.Update(ctx, account); err != nil {
			return err
		}

		refID := w.ID
		t, err := domain.NewTransaction(
			u.newID(), accountID, -w.Amount, domain.TransactionWishPaid, &refID, occurredOn,
		)
		if err != nil {
			return err
		}
		if err := u.transactions.Create(ctx, t); err != nil {
			return err
		}

		paid = w
		return nil
	})
	if err != nil {
		return domain.Wish{}, err
	}
	return paid, nil
}

func (u *WishUsecase) Delete(ctx context.Context, id uuid.UUID) error {
	if _, err := u.wishes.Get(ctx, id); err != nil {
		return err
	}
	return u.wishes.Delete(ctx, id)
}

File: worker/src/usecase/wish.ts
import { negateMoney, type Money } from '../domain/money'
import type { IsoDate } from '../domain/time'
import { Transaction } from '../domain/transaction'
import { Wish, type WishCategory, type WishStatus } from '../domain/wish'
import type {
  AccountRepository,
  AtomicWriter,
  Clock,
  IDGenerator,
  WishRepository,
  WriteOperation,
} from './port'

/** PATCH の部分更新を表す。undefined の項目は変更しない。 */
export type UpdateWishInput = {
  title?: string
  amount?: Money
  category?: WishCategory
  priority?: number
  /** 期限。null を渡すと期限を外す。undefined は「変更しない」。 */
  deadline?: IsoDate | null
}
// status が無いのは、状態を動かせるのが commit / pay / drop だけのため（不変条件6）。

export class WishUsecase {
  readonly #writer: AtomicWriter
  readonly #wishes: WishRepository
  readonly #accounts: AccountRepository
  readonly #now: Clock
  readonly #newID: IDGenerator

  constructor(
    writer: AtomicWriter,
    wishes: WishRepository,
    accounts: AccountRepository,
    now: Clock,
    newID: IDGenerator,
  ) {
    this.#writer = writer
    this.#wishes = wishes
    this.#accounts = accounts
    this.#now = now
    this.#newID = newID
  }

  list(status: WishStatus | null): Promise<Wish[]> {
    return this.#wishes.list(status)
  }

  async create(
    title: string,
    amount: Money,
    category: WishCategory,
    priority: number,
    deadline: IsoDate | null,
  ): Promise<Wish> {
    const w = Wish.create(this.#newID(), { title, amount, category, priority, deadline })
    await this.#wishes.create(w)
    return w
  }

  /** 内容を更新する。状態は動かさない。検証は domain の updateContent が持つ。 */
  async updateContent(id: string, input: UpdateWishInput): Promise<Wish> {
    const w = await this.#wishes.get(id)
    w.updateContent({
      title: input.title ?? w.title,
      amount: input.amount ?? w.amount,
      category: input.category ?? w.category,
      priority: input.priority ?? w.priority,
      // null は「期限を外す」、undefined は「変更しない」。
      deadline: input.deadline === undefined ? w.deadline : input.deadline,
    })
    await this.#wishes.updateContent(w)
    return w
  }

  /** 検討中 → 確定。確定した時点で実質資産から控除される（不変条件3）。 */
  commit(id: string): Promise<Wish> {
    return this.#transit(id, (w) => w.commit())
  }

  /** 検討中 または 確定 → 見送り。 */
  drop(id: string): Promise<Wish> {
    return this.#transit(id, (w) => w.drop())
  }

  /**
   * 遷移して良いかの判定は domain の Wish が持つ。usecase は「どの遷移を
   * 起こしたいか」だけを知る（不変条件6）。
   *
   * 読み取り時の状態を条件に書き込むため、2つのタブから同時に確定しても
   * 2度目は競合として弾かれる。
   */
  async #transit(id: string, transition: (w: Wish) => void): Promise<Wish> {
    const w = await this.#wishes.get(id)
    const expectedStatus = w.status
    transition(w)
    await this.#writer.writeAll([{ kind: 'updateWishStatus', wish: w, expectedStatus }])
    return w
  }

  /**
   * 確定 → 完了に遷移させ、口座残高を減らして履歴を残す。
   *
   * 支払い後、そのウィッシュは確定支出から外れ、同額だけ残高が減る。
   * **実質資産は支払いの前後で変わらない。** これが正しい挙動になる。
   */
  async pay(id: string, accountId: string, occurredOn: IsoDate): Promise<Wish> {
    const wish = await this.#wishes.get(id)
    const expectedStatus = wish.status
    wish.pay()

    const account = await this.#accounts.get(accountId)
    const expectedBalance = account.balance
    account.applyDelta(negateMoney(wish.amount), this.#now())

    const ops: WriteOperation[] = [
      { kind: 'updateWishStatus', wish, expectedStatus },
      { kind: 'updateAccount', account, expectedBalance },
      {
        kind: 'createTransaction',
        transaction: Transaction.create(
          this.#newID(),
          accountId,
          negateMoney(wish.amount),
          'wish_paid',
          wish.id,
          occurredOn,
          // メモは手入力の明細のためのもの。参照先をたどれば何の支払いか分かる。
          '',
        ),
      },
    ]
    await this.#writer.writeAll(ops)
    return wish
  }

  async delete(id: string): Promise<void> {
    await this.#wishes.get(id)
    await this.#wishes.delete(id)
  }
}
